tracer/src/Datadog.Tracer.Native/debugger_rejit_preprocessor.cpp

```diff
@@ -250,9 +253,152 @@ void DebuggerRejitPreprocessor::UpdateMethod(RejitHandlerModuleMethod* methodHan
+            const COR_SIGNATURE* moveNextSig{};
+            hr = metadataImport->FindMethod(nestedAsyncClassOrStruct, WStr("MoveNext"), moveNextSig,
+                                            0, &moveNextMethod);
+            if (FAILED(hr))
+            {
+                Logger::Error("DebuggerRejitPreprocessor::GetMoveNextMethodFromKickOffMethod: failed to call metadataImport->FindMethod for MoveNext method");
+                return  hr;
+            }
+        }
+
+        if (moveNextMethod != mdMethodDefNil)
+        {
+            // we found the correct nested type and MoveNext method
+            break;
+        }
+    }
+    return S_OK;
+}
+
+std::tuple<HRESULT, mdMethodDef, FunctionInfo> DebuggerRejitPreprocessor::TransformKickOffToMoveNext(
+                                                    const ComPtr<IMetaDataImport2>& metadataImport, const ComPtr<IMetaDataEmit2>& metadataEmit,
+                                                    mdMethodDef moveNextMethod, mdTypeDef nestedAsyncClassOrStruct)
+{
+    // this is an async method and we found the generated nested state machine type,
+    // define a new boolean field in the state machine object to indicate whether we have already entered the MoveNext method
+    // (if we have, it means we are re-entering the method as a continuation in a subsequent `await` operation, 
+    // and should not capture the method parameter values as we do the first time around).
+    constexpr BYTE fieldSignature[] = {IMAGE_CEE_CS_CALLCONV_FIELD, ELEMENT_TYPE_BOOLEAN};
+    mdFieldDef isFirstEntry = mdFieldDefNil;
+    auto hr = metadataEmit->DefineField(nestedAsyncClassOrStruct, managed_profiler_debugger_is_first_entry_field_name.c_str(), fdPrivate, fieldSignature,
+                                   sizeof(fieldSignature), 0, nullptr, 0, &isFirstEntry);
+    if (FAILED(hr))
+    {
+        Logger::Error("DebuggerRejitPreprocessor::TransformKickOffToMoveNext: DefineField _isFirstEntry failed");
+        return {hr, mdMethodDefNil, FunctionInfo()};
+    }
+
+    // save the MoveNext method and create a function info for it
+    auto caller = GetFunctionInfo(metadataImport, moveNextMethod);
+    if (!caller.IsValid())
+    {
+        Logger::Error("DebuggerRejitPreprocessor::TransformKickOffToMoveNext: The caller for the methoddef: ",
+                      shared::TokenStr(&moveNextMethod), " is not valid!");
+        return {E_FAIL, mdMethodDefNil, FunctionInfo()};
+    }
+
+    hr = caller.method_signature.TryParse();
+    if (FAILED(hr))
+    {
+        Logger::Error("DebuggerRejitPreprocessor::TransformKickOffToMoveNext: The method signature: ",
+                      caller.method_signature.str(), " cannot be parsed.");
+        return {hr, mdMethodDefNil, FunctionInfo()};
+    }
+
+    return {S_OK, moveNextMethod, FunctionInfo(caller)};
+}
+
+std::tuple<HRESULT, mdMethodDef, FunctionInfo> DebuggerRejitPreprocessor::TransformMethodToRejit(const ComPtr<IMetaDataImport2>& metadataImport,
```

Comment:
Not sure this method name makes sense to me. How about `RejitAsyncMethod`?

Reply:
Fixed in 5a103fbe57cedbc9e13568d4372c82a921c46e26